packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+
+        return tree;
+    }
+
+    // Bar::Baz => <PackageRegistry>::Foo_Package::Bar::Baz
+    ast::TreePtr prependInternalPackageNameToScope(ast::TreePtr scope) {
+        ENFORCE(info != nullptr);
+        // For `Bar::Baz::Bat`, `UnresolvedConstantLit` will contain `Bar`.
+        ast::UnresolvedConstantLit *lastConstLit = ast::cast_tree<ast::UnresolvedConstantLit>(scope);
+        if (lastConstLit != nullptr) {
+            while (auto constLit = ast::cast_tree<ast::UnresolvedConstantLit>(lastConstLit->scope)) {
+                lastConstLit = constLit;
+            }
+        }
+
+        // If `lastConstLit` is `nullptr`, then `scope` should be EmptyTree.
+        ENFORCE(lastConstLit != nullptr || ast::cast_tree<ast::EmptyTree>(scope) != nullptr);
+
+        auto scopeToPrepend =
+            name2Expr(this->info->name.mangledName, name2Expr(core::Names::Constants::PackageRegistry()));
+        if (lastConstLit == nullptr) {
+            return scopeToPrepend;
+        } else {
+            lastConstLit->scope = move(scopeToPrepend);
+            return scope;
+        }
+    }
+
+    // Generates `exportModule`, which dependent packages copy to set up their namespaces.
+    // For package Foo::Bar:
+    //   module Foo::Bar
+    //     ExportedItem1 = <PackageRegistry>::Foo_Bar_Package::Path::To::ExportedItem1
+    //     extend <PackageRegistry>::Foo_Bar_Package::<PackageMethods>
+    //   end
+    void finalize(core::MutableContext ctx) {
+        if (info == nullptr) {
+            if (auto e = ctx.beginError(firstLineOfFile(ctx), core::errors::Packager::InvalidPackageDefinition)) {
+                e.setHeader("Package file must contain a package definition of form `Foo::Bar < PackageSpec`");
+            }
+            return;
+        }
+
+        // NOTE: Don't assign locs to the nodes generated below. They will be copied into other __package.rbs that
+        // import this package with __package.rb-specific locs.
+        for (auto &klass : exported) {
+            // Item = <PackageRegistry>::MangledPackageName::Path::To::Item
+            info->exportedItems.emplace_back(
+                ast::MK::Assign(core::LocOffsets::none(), name2Expr(klass->cnst),
+                                name2Expr(klass->cnst, prependInternalPackageNameToScope(klass->scope.deepCopy()))));
+        }
+
+        if (!exportedMethods.empty()) {
+            // extend <PackageRegistry>::Name_of_package::<PackageMethods>
+            info->exportedItems.emplace_back(
+                ast::MK::Send1(core::LocOffsets::none(), ast::MK::Self(core::LocOffsets::none()), core::Names::extend(),
+                               name2Expr(core::Names::Constants::PackageMethods(),
+                                         name2Expr(this->info->name.mangledName,
+                                                   name2Expr(core::Names::Constants::PackageRegistry())))));
+        }
+    }
+
+    /* Forbidden control flow in packages */
```

Comment:
It's still pretty easy to add flow control with method calls and blocks.

```ruby
loop do
  puts 1
end
```

It seems like this could be easily modified with a custom helper that loops a finite number of times instead of infinitely.

Maybe you want to outlaw blocks too?

Reply:
👍  Yeah, I wasn't sure if the class body itself was a `Block` or not.

Reply:
class body is `vector<ast::Expression>`

method body is `ast::InsSeq` which is a ruby `begin ... end`.

Reply:
Done; blocks and insseq are no longer allowed.